### PR TITLE
[Content] STFT Overview

### PR DESCRIPTION
**Pull request by jamesb93**
Where to look: `/overviews/stft`

### CODE REVIEW

**Review by tremblap — CHANGES_REQUESTED** [review]
Again I recognise material. I think we should use material like I did in Utrecht (as well with what is there) to show more graphically the uncertainty (large window having both a snare and a bass drum is always impactful) and the linearity issues (drawing in octaves in lin, and annotating with nbBinsperOctave in a log graph) - musicians 'get' that. we could even play smearing examples.

the section of phase vocoding we could probably remove I think. otherwise I think the rest has everything we need

File: src/routes/overviews/stft.svx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+### Uncertainty
+The longer the window we take, the finer the frequency-grid we get from the DFT. This means that there is a trade-off between temporal resolution and frequency resolution. A long window means we are (sort of) averaging information about the signal over a longer time, but getting more detailed information about the spectrum in return. Correspondingly, short windows give us a better impression of moment-to-moment dynamics, but a rougher idea about frequency.
+
+### FFT Size
+Note that the size of the DFT doesn't have to be the same as the window size (but it can't be smaller). In fact, we are normally constrained to use a power of two DFT size because we almost always use an efficient DFT algorithm (the Fast Fourier Transform, FFT) that has this constraint. Sometimes using a bigger DFT / FFT size than the window can be beneficial because it gives us high quality interpolation across frequency for the 'extra' bins we've added.
+
+Note that this is not the same as getting a more precise analysis from the same window: the constraint on how close together in frequency things can be distinguished is governed still by the size of the window we're analysing.
```
Comment (weefuzzy) [inline src/routes/overviews/stft.svx:37]: 'Note that' starts two paragraphs.

**Comment by weefuzzy** [comment]
Not sure I'm up for getting rid of the PV section, given that its associated mental model is a source of much wishful thinking about what STFTs can achieve. If we want to explain that the assumption that a Fourier domain representation can only be conceived in terms of components that have analysable frequencies under some specific conditions, how else would we do it?

File: src/routes/overviews/stft.svx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+If we assume that from window to window, each bin contains some coherent and continuing bit of signal, we can trace the changes in the phase between windows to estimate the frequency of this (hypothetical) component. This model works quite (remarkably) well in a lot of cases, but the naïve assumptions about inferring component frequencies from the phase can lead to familiar artefacts like softened transients, and a chorus-y sound. By and large, if the sound we analyse is well represented by the Phase Vocoder assumptions:
+
+- mostly tonal components
+- that change slowly with respect to the hop size
+- well spaced in frequency with respect to the DFT resolution
+
+As a rule of thumb, if you want to get reliable tracking of partials in a harmonic(ish) sound, you will want to have a conservative 4 frequency bins between partials (because energy gets smooshed across bins in practice). For instance, to reliably track partials in a 100 Hz signal, this implies a bin resolution of 25 Hz. At 44100Hz sampling rate, this in turn implies a window size / FFT size of 2048 samples (44100/2048 = 21.53Hz).
+
+# Resources
```
[the remark this answers is not in the document]
Comment (jamesb93) [reply src/routes/overviews/stft.svx:57]: Yeah, I like your approach in the original learn with "cards". I'm just a lazy bones and haven't gotten around to making a component that I think looks good and communicates well. I've made an issue internally to do so.

[the remark this answers is not in the document]
Comment (jamesb93) [reply src/routes/overviews/stft.svx:57]: Does this look good to you @weefuzzy https://github.com/flucoma/learn-website/pull/158 (check the chroma reference on this PR).

**Comment by tedmoore** [comment]
I am not sure who the intended reader of this is. In general I think it's too jargon-y. There are assumptions being made that I'm not sure we can or should make. I think a chat about the intended readership and the goals of this page would help me get a better sense of what to be considering as I read.

**Comment by jamesb93** [comment]
> I am not sure who the intended reader of this is. In general I think it's too jargon-y. There are assumptions being made that I'm not sure we can or should make. I think a chat about the intended readership and the goals of this page would help me get a better sense of what to be considering as I read.

I think I should remind all reviewers that this is old learn content extracted from Owen's initial putting together of the learn platform. Obviously what we (the new additions) know, the team and almost every circumstance of the project has changed since then, so these have been put up as a sort of stimulus for us to discuss what we might envisage an STFT article to look like now. Certainly our collective attempts to explain the STFT in workshops should influence this (PAs explanation was good and draws on easily understandable examples + metaphors), as well what how we think an STFT primer will support the entangled sprawl of resources we offer. In that light I think we should focus on what we think is valuable, how it can be extended, and what we think can be either removed or perhaps further distilled. I'm not convinced its that jargony if we were to aim it somewhere in the middleground rather than at an absolute beginner. Perhaps glossing over certain concepts (like the distinction between a DFT/FFT) might be valuable to focus the discussion on the way this particular decomposition model influences other things (pitch analysis and minimum frequencies for example). Again, we don't want to reinvent the wheel as there are already tonnes of good splainers out there, like the Jack Schaedler one for example.

**Comment by weefuzzy** [comment]
Thanks for the shielding @jamesb93 😆 

To answer Ted's question, all the techno arcana was there because I was trying to address the problem (or maybe not-problem) that folk in our field 'know' stuff about spectral processing that they've digested by rote, and not knowing where it comes from (I think) gets in the way of being able to properly screw around with stuff. E.g why the powers-of-two thing? It's a condition of the FFTs algorithm design, nothing to do with DFTs. Why distinguish between an STFT and the rest of it? Because that's where challenges with window leakage, overlap add, space monkeys and so forth come in. 

I guess I was pitching for something that bridges the lovely, but still over simplified account in Wishart's _Audible Design_, and the equally lovely but starting to get a bit techy account in Lyon's _Understanding DSP_, or even the Circles, Sines and Signals site.

**Comment by jamesb93** [comment]
> Thanks for the shielding @jamesb93 😆

I know you're a big boy but it is probably non-obvious to Ted the origin of this material which is useful context.

> To answer Ted's question, all the techno arcana was there because I was trying to address the problem (or maybe not-problem) that folk in our field 'know' stuff about spectral processing that they've digested by rote, and not knowing where it comes from (I think) gets in the way of being able to properly screw around with stuff. E.g why the powers-of-two thing? It's a condition of the FFTs algorithm design, nothing to do with DFTs. Why distinguish between an STFT and the rest of it? Because that's where challenges with window leakage, overlap add, space monkeys and so forth come in.
> 
> I guess I was pitching for something that bridges the lovely, but still over simplified account in Wishart's _Audible Design_, and the equally lovely but starting to get a bit techy account in Lyon's _Understanding DSP_, or even the Circles, Sines and Signals site.

Maybe something useful to draw up would be the take home messages we want to instill in this article using those external resources to cover our own gaps and to focus on the things that are perhaps most relevant to our objects (and relying on them as ways of explaining things too).

**Comment by tedmoore** [comment]
Cool. I do like the idea of having a page that gets at those myths and things we learned by rote. I think it could even be titled that way. "Things often misunderstood about FFT". Or something less negative perhaps.

**Comment by jamesb93** [comment]
I set my window to something other than a power of two you won’t guess what happened next?

**Comment by tremblap** [comment]
I think we can unfold. Our discussion about magnitude shows that musicians arrive from a practical perspective with a few assumptions and functional understanding which we can complement softly and not pratronisingly.

section 1: the essential (what i explained in Utrecht)

section 2: a few myth busters (what is in there now)

section 3: further reading for the math savvy (with links to sexy explanations)

(btw @tedmoore I'm still looking for my Lyon book - @jamesb93 I didn't lend it to you by any chance?)

**Pull request closed without merging**